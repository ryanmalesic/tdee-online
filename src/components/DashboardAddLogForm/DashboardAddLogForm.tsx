import { useFormik } from 'formik';
import React from 'react';
import * as Yup from 'yup';

export interface DashboardAddLogFormProps {
  error: string | null;
  loading: boolean;
  onSubmit: (values: DashboardAddLogFormState) => void;
}

export interface DashboardAddLogFormState {
  date: string;
  weight: number;
  caloricIntake: number;
}

const initialValues: DashboardAddLogFormState = {
  date: new Date().toISOString().slice(0, 10),
  weight: 0,
  caloricIntake: 0
};

const DashboardAddLogForm: React.FC<DashboardAddLogFormProps> = (props) => {
  const { error, loading, onSubmit } = props;

  const validationSchema = Yup.object<DashboardAddLogFormState>({
    date: Yup.string()
      .test('date-valid', 'Date must be in the format yyyy-mm-dd', (value) => {
        if (!value) {
          return false;
        }

        const regEx = /^\d{4}-\d{2}-\d{2}$/;

        if (!value.match(regEx)) {
          return false;
        }

        const date = new Date(value);
        const dateTime = date.getTime();

        if (!dateTime && dateTime !== 0) {
          return false;
        }

        return value === date.toISOString().slice(0, 10);
      })
      .required('Date is required'),
    weight: Yup.number().positive('Weight must be a positive value').required('Weight is required'),
    caloricIntake: Yup.number()
      .integer('Caloric intake must be an integer value')
      .min(0, 'Caloric intake must greater than or equal to 0')
      .required('Caloric intake is required')
  });

  const formik = useFormik({ initialValues, validationSchema, onSubmit });

  const submitButtonClass = `button ${error ? 'is-danger' : 'has-background-primary-dark'} ${
    loading ? 'is-loading' : ''
  }`;

  return (
    <form onSubmit={formik.handleSubmit}>
      <div className="columns">
        <div className="column">
          <div className="field">
            <label className="label" htmlFor="date">
              Date
            </label>
            <div className="control">
              {/* eslint-disable-next-line jsx-a11y/autocomplete-valid */}
              <input
                className="input"
                id="date"
                name="date"
                placeholder="yyyy-mm-dd"
                type="date"
                value={formik.values.date}
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
              />
            </div>
            {formik.errors.date && formik.touched.date && (
              <p className="help is-danger">{formik.errors.date}</p>
            )}
          </div>
        </div>

        <div className="column">
          <div className="field">
            <label className="label" htmlFor="weight">
              Weight
            </label>
            <div className="control">
              <input
                className="input"
                id="weight"
                name="weight"
                type="number"
                value={formik.values.weight}
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
              />
            </div>
            {formik.errors.weight && formik.touched.weight && (
              <p className="help is-danger">{formik.errors.weight}</p>
            )}
          </div>
        </div>

        <div className="column">
          <div className="field">
            <label className="label" htmlFor="caloricIntake">
              Caloric Intake
            </label>
            <div className="control">
              <input
                className="input"
                id="caloricIntake"
                name="caloricIntake"
                type="number"
                value={formik.values.caloricIntake}
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
              />
            </div>
            {formik.errors.caloricIntake && formik.touched.caloricIntake && (
              <p className="help is-danger">{formik.errors.caloricIntake}</p>
            )}
          </div>
        </div>
      </div>

      <div className="field ">
        <div className="control">
          <button className={submitButtonClass} disabled={formik.isSubmitting} type="submit">
            Add log
          </button>
        </div>
      </div>

      {error && <p className="help is-danger">{error}</p>}
    </form>
  );
};

export default DashboardAddLogForm;
