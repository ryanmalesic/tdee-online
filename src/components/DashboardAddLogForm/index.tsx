import DashboardAddLogForm, { DashboardAddLogFormState } from './DashboardAddLogForm';

export { DashboardAddLogForm as default };
export type { DashboardAddLogFormState };
