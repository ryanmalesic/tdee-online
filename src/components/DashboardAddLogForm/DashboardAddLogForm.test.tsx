import React from 'react';
import renderer from 'react-test-renderer';

import DashboardAddLogForm from './DashboardAddLogForm';

it('renders DashboardAddLogForm unchanged', () => {
  const tree = renderer.create(<DashboardAddLogForm />).toJSON();
  expect(tree).toMatchSnapshot();
});
