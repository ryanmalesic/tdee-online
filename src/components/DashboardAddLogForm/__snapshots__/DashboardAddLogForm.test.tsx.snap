// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DashboardAddLogForm unchanged 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="columns"
  >
    <div
      className="column"
    >
      <div
        className="field"
      >
        <label
          className="label"
          htmlFor="date"
        >
          Date
        </label>
        <div
          className="control"
        >
          <input
            className="input"
            id="date"
            name="date"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="yyyy-mm-dd"
            type="date"
            value="2020-09-07"
          />
        </div>
      </div>
    </div>
    <div
      className="column"
    >
      <div
        className="field"
      >
        <label
          className="label"
          htmlFor="weight"
        >
          Weight
        </label>
        <div
          className="control"
        >
          <input
            className="input"
            id="weight"
            name="weight"
            onBlur={[Function]}
            onChange={[Function]}
            step="0.1"
            type="number"
            value={0}
          />
        </div>
      </div>
    </div>
    <div
      className="column"
    >
      <div
        className="field"
      >
        <label
          className="label"
          htmlFor="caloricIntake"
        >
          Caloric Intake
        </label>
        <div
          className="control"
        >
          <input
            className="input"
            id="caloricIntake"
            name="caloricIntake"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value={0}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="field "
  >
    <div
      className="control"
    >
      <button
        className="button has-background-primary-dark "
        disabled={false}
        type="submit"
      >
        Add log
      </button>
    </div>
  </div>
</form>
`;
